Kernel/FileSystem/ISO9660FileSystem.cpp

```diff
@@ -0,0 +1,695 @@
+        get_header();
+    }
+
+    ISO::DirectoryRecordHeader const* operator*() { return m_current_header; }
+
+    // Recurses into subdirectories. May fail.
+    KResultOr<bool> next()
+    {
+        if (done())
+            return false;
+        dbgln_if(ISO9660_VERY_DEBUG, "next(): Called");
+
+        if (has_flag(m_current_header->file_flags, ISO::FileFlags::Directory)) {
+            dbgln_if(ISO9660_VERY_DEBUG, "next(): Recursing");
+            {
+                bool result = m_directory_stack.try_append(move(m_current_directory));
+                if (!result) {
+                    return ENOMEM;
+                }
+            }
+
+            dbgln_if(ISO9660_VERY_DEBUG, "next(): Pushed into directory stack");
+
+            {
+                auto result = read_directory_contents();
+                if (result.is_error()) {
+                    return result;
+                }
+            }
+
+            dbgln_if(ISO9660_VERY_DEBUG, "next(): Read directory contents");
+
+            m_current_directory.offset = 0;
+            get_header();
+            if (m_current_header->length == 0) {
+                // We have found an empty directory, let's continue with the
+                // next one.
+                if (!go_up())
+                    return false;
+            } else {
+                // We cannot skip here, as this is the first record in this
+                // extent.
+                return true;
+            }
+        }
+
+        return skip();
+    }
+
+    // Skips to the directory in the list, returns whether there was a next one.
+    // No allocation here, cannot fail.
+    bool skip()
+    {
+        VERIFY(m_current_directory.entry);
+
+        if (m_current_directory.offset >= m_current_directory.entry->length) {
+            dbgln_if(ISO9660_VERY_DEBUG, "skip(): Was at last item already");
+            return false;
+        }
+
+        m_current_directory.offset += m_current_header->length;
+        get_header();
+        if (m_current_header->length == 0) {
+            // According to ECMA 119, if a logical block contains directory
+            // records, then the leftover bytes in the logical block are
+            // all zeros. So if our directory header has a length of 0,
+            // we're probably looking at padding.
+            //
+            // Of course, this doesn't mean we're done; it only means that there
+            // are no more directory entries in *this* logical block.  If we
+            // have at least one more logical block of data length to go, we
+            // need to snap to the next logical block, because directory records
+            // cannot span multiple logical blocks.
+            u32 remaining_bytes = m_current_directory.entry->length - m_current_directory.offset;
+            if (remaining_bytes > m_fs.logical_block_size()) {
+                m_current_directory.offset += remaining_bytes % m_fs.logical_block_size();
```

Comment:
should we be using checked math for all these calculations? What if an attacker asks us to mount a "comprosmised" ISO?

Reply:
In a few places that can be useful, but most of the actual data access is taken care of by `BlockBasedFileSystem` which does report errors on OOB access. I'll see whether I need to sprinkle some checked math.

Reply:
I've looked, and it seems like anything that would cause an over/underflow would be caught by `raw_read` failing, so I think it's fine as it is.